src/middlewares/snapshot-middleware.ts

```diff
@@ -20,43 +20,88 @@ export class SnapshotMiddleware {
             info: any,
         ) => {
             this.logger.debug('Snapshot middleware started job', context);
-            let startIndex;
-            let countPerPage;
+            let startIndex: number;
+            let countPerPage: number;
+            let currentPage: any[];
             const result = await resolve(root, args, context, info);
+
             if (!(result && result.totalCount && result.getData && !root)) {
                 return result;
             }
 
             if (context.requestHeaders.snapRequest || this.snapshotConfiguration.autoSnapshot) {
                 if (context.snapshotContext == null) {
-                    const totalCount = await result.totalCount();
-
-                    countPerPage = context.requestHeaders.snapPageSize
-                        ? Math.min(
-                              this.snapshotConfiguration.maxPageSize,
-                              context.requestHeaders.snapPageSize,
-                          )
-                        : this.snapshotConfiguration.maxPageSize;
-
-                    if (this.snapshotConfiguration.autoSnapshot) {
-                        if (totalCount > countPerPage) {
-                            context.returnedExtensions.totalCount = totalCount;
-                        } else {
-                            countPerPage = totalCount;
-                        }
-                    } else if (context.requestHeaders.snapRequest) {
-                        context.returnedExtensions.totalCount = totalCount;
-                    }
-
+                    countPerPage = await this.calculateSnapshotParameters(result, context);
                     startIndex = 0;
                 } else {
                     startIndex = context.snapshotContext.startIndex!;
                     countPerPage = context.snapshotContext.countPerPage!;
                 }
+
+                currentPage = await this.prefetchEntities(
+                    context,
+                    result,
+                    startIndex,
+                    countPerPage,
+                );
+            } else {
+                currentPage = await result.getData(0, undefined);
             }
 
             this.logger.debug('Snapshot middleware finished job', context);
-            return result.getData(startIndex, countPerPage);
+            return currentPage;
         };
     }
+
+    private async calculateSnapshotParameters(result: any, context: PolarisGraphQLContext) {
```

Comment:
i think tou can use a better name for this method. the name isn't describes the best what this method is doing.